crypto/params.c

```diff
@@ -0,0 +1,254 @@
+/*
+ * Copyright 2019 The OpenSSL Project Authors. All Rights Reserved.
+ * Copyright (c) 2019, Oracle and/or its affiliates.  All rights reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <string.h>
+#include <openssl/bn.h>
+#include <openssl/params.h>
+#include <openssl/err.h>
+#include <openssl/cryptoerr.h>
+
+OSSL_PARAM *OSSL_PARAM_locate(const OSSL_PARAM *p, const char *key)
+{
+    for (; p->key != NULL; p++)
+        if (strcmp(key, p->key) == 0)
+            return (OSSL_PARAM *)p;
+    return NULL;
+}
+
+static size_t OSSL_PARAM_get_width(int type)
+{
+    switch (type) {
+    case OSSL_PARAM_INT:        return sizeof(int);
+    case OSSL_PARAM_UINT:       return sizeof(unsigned int);
+    case OSSL_PARAM_INT64:      return sizeof(int64_t);
+    case OSSL_PARAM_UINT64:     return sizeof(uint64_t);
+    case OSSL_PARAM_LONG:       return sizeof(long);
+    case OSSL_PARAM_ULONG:      return sizeof(unsigned long);
+    case OSSL_PARAM_SIZET:      return sizeof(size_t);
+    case OSSL_PARAM_DOUBLE:     return sizeof(double);
+    case OSSL_PARAM_POINTER:    return sizeof(void *);
+    }
+    return -1;
```

Comment:
This returns a size_t, so returning -1 on error doesn't make a lot of sense. Probably 0 is better.